pixy/consumer.go

```diff
@@ -0,0 +1,562 @@
+type groupConsumer struct {
+	baseCID               *sarama.ContextID
+	config                *Config
+	group                 string
+	sc                    *SmartConsumer
+	dispatcher            *dispatcher
+	kafkaClient           sarama.Client
+	dumbConsumer          sarama.Consumer
+	offsetMgr             sarama.OffsetManager
+	kazooConn             *kazoo.Kazoo
+	registry              *consumerGroupRegistry
+	exclusiveConsumers    map[string]map[int32]*exclusiveConsumer
+	addTopicConsumerCh    chan *topicConsumer
+	deleteTopicConsumerCh chan *topicConsumer
+	stoppingCh            chan none
+	wg                    sync.WaitGroup
+}
+
+func (sc *SmartConsumer) newConsumerGroup(group string) *groupConsumer {
+	gc := &groupConsumer{
+		baseCID:               sc.baseCID.NewChild(group),
+		config:                sc.config,
+		group:                 group,
+		sc:                    sc,
+		kafkaClient:           sc.kafkaClient,
+		dumbConsumer:          sc.dumbConsumer,
+		offsetMgr:             sc.offsetMgr,
+		kazooConn:             sc.kazooConn,
+		exclusiveConsumers:    make(map[string]map[int32]*exclusiveConsumer),
+		addTopicConsumerCh:    make(chan *topicConsumer),
+		deleteTopicConsumerCh: make(chan *topicConsumer),
+		stoppingCh:            make(chan none),
+	}
+	gc.dispatcher = newDispatcher(gc.baseCID, gc, sc.config)
+	return gc
+}
+
+func (gc *groupConsumer) String() string {
+	return gc.baseCID.String()
+}
+
+func (gc *groupConsumer) addTopicConsumer() chan<- *topicConsumer {
+	return gc.addTopicConsumerCh
+}
+
+func (gc *groupConsumer) deleteTopicConsumer() chan<- *topicConsumer {
+	return gc.deleteTopicConsumerCh
+}
+
+func (gc *groupConsumer) key() string {
+	return gc.group
+}
+
+func (gc *groupConsumer) start(stoppedCh chan<- dispatchTier) {
+	spawn(&gc.wg, func() {
+		defer func() { stoppedCh <- gc }()
+		gc.registry = spawnConsumerGroupRegister(gc.group, gc.config.ClientID, gc.config, gc.kazooConn)
+		var manageWg sync.WaitGroup
+		spawn(&manageWg, gc.managePartitions)
+		gc.dispatcher.start()
+		<-gc.stoppingCh
+		gc.dispatcher.stop()
+		gc.registry.stop()
+		manageWg.Wait()
+	})
+}
+
+func (gc *groupConsumer) requests() chan<- consumeRequest {
+	return gc.dispatcher.requests()
+}
+
+func (gc *groupConsumer) stop() {
+	close(gc.stoppingCh)
+	gc.wg.Wait()
+}
+
+func (gc *groupConsumer) dispatchKey(req consumeRequest) string {
+	return req.topic
+}
+
+func (gc *groupConsumer) newDispatchTier(key string) dispatchTier {
+	tc := gc.newTopicConsumer(key)
+	return tc
+}
+
+func (gc *groupConsumer) managePartitions() {
+	cid := gc.baseCID.NewChild("managePartitions")
+	defer cid.LogScope()()
+
+	topicConsumers := make(map[string]*topicConsumer)
+	var topics []string
+	var memberSubscriptions map[string][]string
+	ok := true
+	var nilOrRetryCh <-chan time.Time
+	var nilOrRegistryTopicsCh chan<- []string
+	shouldRebalance, canRebalance := false, true
+	rebalanceResultCh := make(chan error, 1)
```

Comment:
I think this would look slightly nicer as a single `var` statement. What do you think?

```go
	var (
		topicConsumers                = make(map[string]*topicConsumer)
		topics                        []string
		memberSubscriptions           map[string][]string
		ok                            = true
		nilOrRetryCh                  <-chan time.Time
		nilOrRegistryTopicsCh         chan<- []string
		shouldRebalance, canRebalance = false, true
		rebalanceResultCh             = make(chan error, 1)
	)
```